### ISSUE TITLE
Still issues compiling 32-bit on Windows 1.9.3

### ISSUE THREAD

**Issue by headius** [headius/ruby-atomic#21]
From #20:

```
Installing atomic (1.1.7) 
Gem::Installer::ExtensionBuildError: ERROR: Failed to build gem native extension.

        C:/ruby/bin/ruby.exe extconf.rb 
creating Makefile

make
generating atomic_reference-i386-mingw32.def
compiling atomic_reference.c
atomic_reference.c: In function 'ir_compare_and_set':
atomic_reference.c:58:2: error: #error No CAS operation available for this platform
make: *** [atomic_reference.o] Error 1


Gem files will remain installed in C:/ruby/lib/ruby/gems/1.9.1/gems/atomic-1.1.7 for inspection.
Results logged to C:/ruby/lib/ruby/gems/1.9.1/gems/atomic-1.1.7/ext/gem_make.out
An error occurred while installing atomic (1.1.7), and Bundler cannot continue.
Make sure that `gem install atomic -v '1.1.7'` succeeds before bundling.
```

Upgrading to 2.0.0 works, but it's not really an appropriate solution.

**Comment by whitequark** [comment]
You need this: http://msdn.microsoft.com/en-us/library/windows/desktop/ms683562(v=vs.85).aspx in a #ifdef WIN32 section.

I'd love to submit a PR but setting up a Windows machine is too much time.

**Comment by whitequark** [comment]
Also, ir_compare_and_set should really be compare_and_swap, as that's how the CAS acronym decodes

**Comment by headius** [comment]
@whitequark Thanks! Hopefully I'll get a Windows dev env set up or lean on someone else who already has it.

**Comment by phasis68** [comment]
The subject of this issue is wrong.
I guess the issuer's OS is 64-bit and ruby 1.9.3 is 32-bit and the compiler is mingw32 gcc 4.5.2.

Anyway, here is a patch:

```
diff --git a/extconf.rb b/extconf.rb.new
index 00f2bb3..6ea5c08 100644
--- a/extconf.rb
+++ b/extconf.rb.new
@@ -14,7 +14,12 @@ require 'mkmf'
 extension_name = 'atomic_reference'
 dir_config(extension_name)

-try_run(<<CODE) && ($defs << '-DHAVE_GCC_CAS')
+case CONFIG["arch"]
+when /mswin32|mingw/
+    $CFLAGS += " -march=native"
+end
+
+try_run(<<CODE,$CFLAGS) && ($defs << '-DHAVE_GCC_CAS')
 int main() {
   int i = 1;
   __sync_bool_compare_and_swap(&i, 1, 4);
@@ -22,9 +27,4 @@ int main() {
 }
 CODE

-case CONFIG["arch"]
-when /mswin32|mingw/
-    $CFLAGS += " -march=native"
-end
-
 create_makefile(extension_name)
```

**Comment by headius** [comment]
@phasis68 You are correct. I updated the title.

And thanks for the patch! I'll does it build on MRI and pass the tests I added?

**Comment by phasis68** [comment]
Here is a patched test result:

```
# Running tests:

.F......

Finished tests in 0.018001s, 444.4198 tests/s, 722.1821 assertions/s.

  1) Failure:
test_numeric_cas(TestAtomic) [C:/work/ruby-atomic-master/test/test_atomic.rb:85]  :
CAS failed for numeric 
115792089237316195423570985008687907853269984665640564039457584007913129639936 => 
115792089237316195423570985008687907853269984665640564039457584007913129639937

8 tests, 13 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by headius** [comment]
Great! So it's working except for the new unimplemented numeric behavior.

I will proceed to patch. Thanks!

**Comment by headius** [comment]
Fixed in 6c2ce925.

atomic-1.1.8 is on rubygems.org.
